### DISCUSSION TITLE
GetCurrentReading has no data and the timestamp is frozen.

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [microsoft/GDK#62] [Q&A]
I'm calling the function from C#. And I just can't get any data out of the reading. GetTimestamp returns the same timestamp every call. And the mouse-state is filled with zero's. The Input interface is returning a valid timestamp it seems. I just can't make any progress on this. Any help is appreciated. I'm just going round in circles.

**CALLING CODE**

    `
    public static void PollMouse()
    {
        if (g_GameInput == null)
            return;

    try
    {
        // get pointer to reading
        uint result = g_GameInput.GetCurrentReading(InputKind.Mouse, IntPtr.Zero, out IntPtr readingPtr);

        // convert int-ptr to IGameInputReading
        object obj = Marshal.GetObjectForIUnknown(readingPtr);
        IInputReading reading = (IInputReading)obj;
        if (reading == null)
            Console.WriteLine("reading is null" + result.ToString());
        else
        {
            try
            {
                if (g_Mouse == null)
                {
                    reading.GetDevice(out IntPtr mouseDevPtr);
                    object mouseDevObj = Marshal.GetObjectForIUnknown(mouseDevPtr);
                    g_Mouse = (IInputDevice)mouseDevObj;
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Input.PollMouse | GetDevice: " + e.ToString());
            }

            ulong g_t = g_GameInput.GetCurrentTimestamp();
            ulong r_t = reading.GetTimestamp();

            MouseState state = new();
            bool bresult = reading.GetMouseState(ref state);
            InputKind kind = reading.GetInputKind();

            string binText = Convert.ToString((int)state.Buttons, 2).PadLeft(8, '0');
            Console.WriteLine(g_t.ToString() + " | " + r_t.ToString() + " | " + result.ToString() + " | " + kind.ToString() + " | " + state.DeltaSumX + "," + state.DeltaSumY + " - " + binText);
        }

    }
    catch (Exception e)
    {
        Console.WriteLine("Input.PollMouse | GetCurrentReading: " + e.ToString());
    }
   `
    
**INPUT CLASS**
    
```
    [ComImport]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    [Guid("11BE2A7E-4254-445A-9C09-FFC40F006918")]
    public interface IGameInput
    {
    // IFACEMETHOD_(uint64_t, GetCurrentTimestamp)(THIS) PURE;
    
    [PreserveSig]
    public ulong GetCurrentTimestamp();

    //    _Must_inspect_result_ IFACEMETHOD(GetCurrentReading)(THIS_
    //_In_ GameInputKind inputKind,
    //    _In_opt_ IGameInputDevice * device,
    //    _COM_Outptr_ IGameInputReading ** reading) PURE;

    [PreserveSig]
    public uint GetCurrentReading(
    InputKind inputKind,
    IntPtr device,
    out IntPtr reading);            // this is correct
```

    
**READING CLASS**
```
    [ComImport]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    [Guid("2156947A-E1FA-4DE0-A30B-D812931DBD8D")]
    public interface IInputReading
    {
    //IFACEMETHOD_(GameInputKind, GetInputKind)(THIS) PURE;

    [PreserveSig]
    public InputKind GetInputKind();

    //IFACEMETHOD_(uint64_t, GetSequenceNumber)(THIS_
    //    _In_ GameInputKind inputKind) PURE;

    public ulong GetSequenceNumber(InputKind inputKind);

    //IFACEMETHOD_(uint64_t, GetTimestamp)(THIS) PURE;

    [PreserveSig]
    public ulong GetTimestamp();

    //IFACEMETHOD_(void, GetDevice)(THIS_
    //    _Outptr_ IGameInputDevice ** device) PURE;

    [PreserveSig]
    public void GetDevice(out IntPtr device);
    
    ...    
    
    //IFACEMETHOD_(bool, GetMouseState)(THIS_
    //    _Out_ GameInputMouseState * state) PURE;
    
    [PreserveSig]
    public bool GetMouseState(ref MouseState state);
```

**Comment by bbowman** [comment]
This looks like a duplicate of https://github.com/microsoft/GDK/issues/64 and a possible bug with GameInput Readings. Closing this discussion in favor of the issue though this is good information so will link to this from there.
